File: 06/assembler.cpp
#include <iostream>
#include <fstream>
#include <algorithm>
#include <string>
#include <vector>
#include <cctype>
#include <cassert>
#include <bitset>
#include <map>


class Parser {
    std::ifstream infile;
    std::vector<std::string> commands;
    std::string currentCommand;
    int idx;

public:
    Parser(std::string filename) : infile(filename) {
        assert(infile.is_open());

        std::string line;
        while (std::getline(infile, line)) {
            // remove whitespace characters
            line.erase(std::remove_if(line.begin(), line.end(), isspace), line.end());
            // remove comments
            int i = line.find("//");
            if (i != std::string::npos) {
                line.erase(line.begin() + i, line.end());
            }

            if (!line.empty()) {
                commands.push_back(line);
            }
        }

        infile.close();
    }

    bool hasMoreCommands() {
        return idx < commands.size();
    }

    void advance() {
        assert(hasMoreCommands());
        currentCommand = commands[idx++];
    }

    std::string commandType() {
        if (currentCommand[0] == '@') {
            return "A_COMMAND";
        } else if (currentCommand[0] == '(') {
            return "L_COMMAND";
        } else {
            return "C_COMMAND";
        }
    }

    std::string symbol() {
        if (commandType() == "A_COMMAND") {
            return currentCommand.substr(1);
        } else {
            return currentCommand.substr(1, currentCommand.size() - 2);
        }
    }

    std::string dest() {
        int i = currentCommand.find('=');
        if (i == std::string::npos) {
            return "null";
        } else {
            return currentCommand.substr(0, i);
        }
    }

    std::string comp() {
        int i = currentCommand.find('=');
        if (i == std::string::npos) {
            i = -1;
        }
        int j = currentCommand.find(';');
        if (j == std::string::npos) {
            j = currentCommand.size();
        }
        return currentCommand.substr(i + 1, j - i - 1);
    }

    std::string jump() {
        int j = currentCommand.find(';');
        if (j == std::string::npos) {
            return "null";
        } else {
            return currentCommand.substr(j + 1);
        }
    }

    void reset() {
        idx = 0;
    }
};


class Code {
    const std::vector<std::string> destCommands = {
        "null", "M", "D", "MD", "A", "AM", "AD", "AMD"
    };

    const std::vector<std::string> jumpCommands = {
        "null", "JGT", "JEQ", "JGE", "JLT", "JNE", "JLE", "JMP"
    };

public:
    int dest(std::string s) {
        return std::find(destCommands.begin(), destCommands.end(), s) - destCommands.begin();
    }

    int comp(std::string s) {
        if (s == "0")   return 0b0101010;
        if (s == "1")   return 0b0111111;
        if (s == "-1")  return 0b0111010;
        if (s == "D")   return 0b0001100;
        if (s == "A")   return 0b0110000;
        if (s == "!D")  return 0b0001101;
        if (s == "!A")  return 0b0110001;
        if (s == "-D")  return 0b0001111;
        if (s == "-A")  return 0b0110011;
        if (s == "D+1") return 0b0011111;
        if (s == "A+1") return 0b0110111;
        if (s == "D-1") return 0b0001110;
        if (s == "A-1") return 0b0110010;
        if (s == "D+A") return 0b0000010;
        if (s == "D-A") return 0b0010011;
        if (s == "A-D") return 0b0000111;
        if (s == "D&A") return 0b0000000;
        if (s == "D|A") return 0b0010101;
        if (s == "M")   return 0b1110000;
        if (s == "!M")  return 0b1110001;
        if (s == "-M")  return 0b1110011;
        if (s == "M+1") return 0b1110111;
        if (s == "M-1") return 0b1110010;
        if (s == "D+M") return 0b1000010;
        if (s == "D-M") return 0b1010011;
        if (s == "M-D") return 0b1000111;
        if (s == "D&M") return 0b1000000;
        if (s == "D|M") return 0b1010101;
        return -1;
    }

    int jump(std::string s) {
        return std::find(jumpCommands.begin(), jumpCommands.end(), s) - jumpCommands.begin();
    }
};


class SymbolTable {
    std::map<std::string, int> table;

public:
    SymbolTable() {
        table["SP"]   = 0;
        table["LCL"]  = 1;
        table["ARG"]  = 2;
        table["THIS"] = 3;
        table["THAT"] = 4;
        for (int i = 0; i < 16; ++i) {
            table["R" + std::to_string(i)] = i;
        }
        table["SCREEN"] = 16384;
        table["KBD"]    = 24576;
    }

    void addEntry(std::string symbol, int address) {
        table[symbol] = address;
    }

    bool contains(std::string symbol) {
        return table.count(symbol);
    }

    int getAddress(std::string symbol) {
        return table[symbol];
    }
};


int main(int argc, char *argv[]) {
    std::string infilename = argv[1];
    std::string outfilename = infilename.substr(0, infilename.size() - 4) + ".hack";
    std::ofstream outfile;
    outfile.open(outfilename);

    Parser parser(infilename);
    Code code;
    SymbolTable table;

    // associate labels with addresses
    int address = 0;
    while (parser.hasMoreCommands()) {
        parser.advance();
        if (parser.commandType() == "L_COMMAND") {
            table.addEntry(parser.symbol(), address);
        } else {
            ++address;
        }
    }

    // convert each command to binary
    parser.reset();
    int varAddress = 16;
    while (parser.hasMoreCommands()) {
        parser.advance();
        auto type = parser.commandType();
        if (type == "A_COMMAND") {
            auto s = parser.symbol();
            int a = 0;
            if (std::isdigit(s[0])) {  // number
                a = std::stoi(s);
            } else if (table.contains(s)) {  // symbol that is already defined
                a = table.getAddress(s);
            } else {  // new symbol
                table.addEntry(s, varAddress);
                a = varAddress;
                ++varAddress;
            }
            std::bitset<16> bin(a);
            outfile << bin << std::endl;
        }
        if (type == "C_COMMAND") {
            int c = code.comp(parser.comp());
            int d = code.dest(parser.dest());
            int j = code.jump(parser.jump());
            std::bitset<16> bin((7 << 13) + (c << 6) + (d << 3) + j);
            outfile << bin << std::endl;
        }
    }

    outfile.close();
    return 0;
}

File: assembler.jl
@enum CommandType a_command l_command c_command


# given an input file, return a list of commands without comments and whitespaces
function readcommands(path::String)::Vector{String}
    commands = []
    open(path, "r") do file
        for line in eachline(file)
            # remove whitespace
            line = filter(x -> !isspace(x), line)

            # remove comments
            res = findfirst("//", line)
            if !isnothing(res)
                i = res[1]
                line = line[1:i-1]
            end

            if !isempty(line)
                push!(commands, line)
            end
        end
    end
    commands
end


# get the type of the given command
function commandtype(command::String)::CommandType
    if command[1] == '@'
        a_command
    elseif command[1] == '('
        l_command
    else
        c_command
    end
end


# get the symbol of the given command
# the command should be of either a_command or l_command
function symbol(command::String)::String
    if commandtype(command) == a_command
        command[2:end]
    else
        command[2:end-1]
    end
end


# get the dest part of the given command
# the command should be of c_command
function dest(command::String)::String
    i = findfirst(isequal('='), command)
    if isnothing(i)
        "null"
    else
        command[1:i-1]
    end
end


# get the comp of the given command
# the command should be of C_COMMAND
function comp(command::String)::String
    i = findfirst(isequal('='), command)
    if isnothing(i)
        i = 0
    end
    j = findfirst(isequal(';'), command)
    if isnothing(j)
        j = length(command) + 1
    end
    command[i+1:j-1]
end


# get the jump part of the given command
# the command should be of C_COMMAND
function jump(command::String)::String
    j = findfirst(isequal(';'), command)
    if isnothing(j)
        "null"
    else
        command[j+1:end]
    end
end


# convert dest to binary
function convertdest(s::String)::UInt16
    bin = 0
    'M' in s && (bin += 1)
    'D' in s && (bin += 2)
    'A' in s && (bin += 4)
    bin
end


# convert comp to binary
function convertcomp(s::String)::UInt16
    s == "0"   && return 0b0101010
    s == "1"   && return 0b0111111
    s == "-1"  && return 0b0111010
    s == "D"   && return 0b0001100
    s == "A"   && return 0b0110000
    s == "!D"  && return 0b0001101
    s == "!A"  && return 0b0110001
    s == "-D"  && return 0b0001111
    s == "-A"  && return 0b0110011
    s == "D+1" && return 0b0011111
    s == "A+1" && return 0b0110111
    s == "D-1" && return 0b0001110
    s == "A-1" && return 0b0110010
    s == "D+A" && return 0b0000010
    s == "D-A" && return 0b0010011
    s == "A-D" && return 0b0000111
    s == "D&A" && return 0b0000000
    s == "D|A" && return 0b0010101
    s == "M"   && return 0b1110000
    s == "!M"  && return 0b1110001
    s == "-M"  && return 0b1110011
    s == "M+1" && return 0b1110111
    s == "M-1" && return 0b1110010
    s == "D+M" && return 0b1000010
    s == "D-M" && return 0b1010011
    s == "M-D" && return 0b1000111
    s == "D&M" && return 0b1000000
    s == "D|M" && return 0b1010101
end


# convert jump to binary
function convertjump(s::String)::UInt16
    findfirst(isequal(s), ["null", "JGT", "JEQ", "JGE", "JLT", "JNE", "JLE", "JMP"]) - 1
end


function main()
    inpath = ARGS[1]
    i = findlast(isequal('.'), inpath)
    outpath = inpath[1:i-1] * ".hack"

    commands = readcommands(inpath)

    # symbol table with predefined names
    symbols = Dict{String,UInt16}(
        "SP" => 0,
        "LCL" => 1,
        "ARG" => 2,
        "THIS" => 3,
        "THAT" => 4,
        "SCREEN" => 16384,
        "KBD" => 24576
    )
    for i in 0:15
        symbols["R" * string(i)] = i
    end

    open(outpath, "w") do file
        # associate labels with addresses
        address = 0
        for command in commands
            if commandtype(command) == l_command
                symbols[symbol(command)] = address
            else
                address += 1
            end
        end

        # convert each command to binary
        address::UInt16 = 16
        for command in commands
            type = commandtype(command)
            type == l_command && continue
            if type == a_command
                s = symbol(command)
                bin = if isdigit(s[1])
                    parse(UInt16, s)
                else
                    if !haskey(symbols, s)
                        symbols[s] = address
                        address += 1
                    end
                    symbols[s]
                end
            else
                c = convertcomp(comp(command))
                d = convertdest(dest(command))
                j = convertjump(jump(command))
                bin = (UInt16(7) << 13) + (c << 6) + (d << 3) + j
            end
            write(file, bitstring(bin) * "\n")
        end
    end
end


main()